lib/api/controllers/debugController.ts

```diff
@@ -0,0 +1,108 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2022 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+import { Request } from '../request';
+import { NativeController } from './baseController';
+import v8 from 'v8';
+import { HttpStream } from '../../types';
+
+/**
+ * @class DebugController
+ */
+export class DebugController extends NativeController {
+  constructor () {
+    super([
+      'heapSnapshot',
+      'getHeapStatistics',
+      'getHeapSpaceStatistics',
+      'getHeapCodeStatistics',
+      'setFlags',
+    ]);
+  }
+
+  /**
+   * Take a heap snapshot and returns the filename of the snapshot.
+   * If the download parameter is set to true, the snapshot will be downloaded using an HTTP Stream instead of saved on the disk.
+   * 
+   * @param {Request} request
+   */
+  async heapSnapshot (request) {
+    if (request.getBoolean('download')) {
+      if (request.context.connection.protocol !== 'http') {
+        throw new Error('Downloading heap snapshots is only supported over HTTP');
+      }
+
+      const stream = v8.getHeapSnapshot();
+
+      const date = new Date();
+      const filename = `heap-${date.getFullYear()}-${date.getMonth()}-${date.getDay()}-${date.getHours()}-${date.getMinutes()}.heapsnapshot`;
+      request.response.configure({
+        headers: {
+          'Content-Disposition': `attachment; filename="${filename}"`,
+          'Content-Type': 'application/json',
+        }
+      });
+
+      return new HttpStream(stream);
+    }
+
+    return v8.writeHeapSnapshot();
```

Comment:
I'm not sure this is very useful since the snapshot will only be saved on a specific node, with no guarantee of being accessible.